### PR TITLE
Support arithmetic operations on numerical fields for scalar filtering

### PR DESCRIPTION
**Pull request by avsolatorio**
This enhancement will allow users to include binary arithmetic operations on fields defined in filter expressions passed into the predicate of a query.

The following examples of expressions are now supported after this patch.

```
expr="(int64 + 10 == 29)"
expr="(int64 - 64 == 2016)"
expr="(int64 * 59 == 1029)"
expr="(int64 / 11 == 2)"
expr="(int64 % 11 == 0)"
expr="(int64 / 11 == 2) || int64 in [1, 2, 3, 4]"
```

Allowing arithmetic operations on fields can allow us to implement complex heuristics based on the encoding of the fields.  For example, if we encode multi-valued fields as products of prime numbers, we can effectively filter these fields using the modulo of the corresponding prime number assigned to the value in the multi-valued field.

Concretely, say we have the data `data1=["Philippines", "Japan", "China", "USA"]` and `data2=["Indonesia", "Malawi", "China", "Philippines"]` which we represent as prime numbers:

```
Philippines -> 2
Japan -> 3
China -> 5
USA -> 7
Indonesia -> 11
Malawi -> 13
```

we encode these data as product of primes `p_data1=210` and `p_data2=2002` and store them in an INT64 field `p_countries`.

Now, if a user wants to find data that contain "Philippines", we can simply take the prime associated with the input (Philippines=2) and perform the expression `(p_countries % 2 == 0)`. This will be true for both `p_data1` and `p_data2`.

Alternatively, if a user wants to find the data that contain "Malawi", we use the prime (Malawi=13) and perform the expression `(p_countries % 13 == 0)`. This will return only `p_data2`.

This method will generally permit users to store multi-valued data, albeit ones with low cardinality (due to the INT64 limit), and filter them within Milvus (no need to fetch batches and filter post query). This is helpful for data with labels that are rare, which are likely to be missed out if we don't filter directly the data.

Signed-off-by: Aivin V. Solatorio <[avsolatorio@gmail.com](mailto:avsolatorio@gmail.com)>
/kind feature
issue: https://github.com/milvus-io/milvus/issues/16513

### CODE REVIEW

**Comment by xiaofan-luan** [comment]
@DragonDriver please help on review this issue?

**Comment by xiaofan-luan** [comment]
hi @avsolatorio , please take a look a look at the go format issue and code coverage~ thanks my man

**Comment by avsolatorio** [comment]
Hello @xiaofan-luan, I have added new test cases to hopefully increase the coverage and I also fixed the go format issue. I believe the code checks and codecov report require manual approval to rerun, right? Thanks!

File: internal/core/src/index/ScalarIndex.h
Comment (DragonDriver) [inline internal/core/src/index/ScalarIndex.h:35]: I don't think `EvalEq` and `EvalNotEq` should be defined in an index. Maybe we should use raw-data to do these two filterings. As you have already implemented, in fact, many index type can't support them.

File: internal/core/src/query/Expr.h
Comment (DragonDriver) [inline internal/core/src/query/Expr.h:116]: `Unk` here is a little confusing. Consider `Unknown` instead.

File: internal/core/src/query/visitors/ExecExprVisitor.cpp
Comment (DragonDriver) [inline internal/core/src/query/visitors/ExecExprVisitor.cpp:245]: As above mentioned, I suggest that we simply use raw data to do brute-force search here. You can abstract a newly function like `ExecRangeVisitorImpl`, this function accepts a `ElemFunc` and will apply the `ElemFunc` for every row in every chunk in the segment.

**Comment by xiaofan-luan** [comment]
> Hello @xiaofan-luan, I have added new test cases to hopefully increase the coverage and I also fixed the go format issue. I believe the code checks and codecov report require manual approval to rerun, right? Thanks!

Sure, I will help on the rerun

File: internal/core/unittest/test_expr.cpp
Comment (DragonDriver) [inline internal/core/unittest/test_expr.cpp:585]: Please also add ut for expr. As for how to generate plan from expr conviniently, maybe you can refer to `plan_parser_test.go` and `translate_text_plan_to_binary_plan`.

File: internal/proto/plan.proto
Comment (DragonDriver) [inline internal/proto/plan.proto:18]: As mentioned above, `Unk` here is a little confusing.

File: internal/proto/plan.proto
Comment (DragonDriver) [inline internal/proto/plan.proto:94]: It seems that `BinaryArithOp` is not used. If so, please delete it.

File: internal/proxy/plan_parser.go
Comment (DragonDriver) [inline internal/proxy/plan_parser.go:78]: What if the `node.Operator` is not of arithmetic operator?

File: internal/proxy/plan_parser.go
Comment (DragonDriver) [inline internal/proxy/plan_parser.go:685]: It's ok that our optimizer handles the ExprTree well (optimizer transfer arithmetic-related node to FunctionNode). You can also consider `BinaryArithCmpExpr` as a special cmp expr and then handle it in cmp expr recursively.
Both these ways make sense for me.

File: internal/core/src/query/visitors/ExecExprVisitor.cpp
Comment (DragonDriver) [inline internal/core/src/query/visitors/ExecExprVisitor.cpp:242]: Besides `==` and `!=`, should we support `<`, `<=`, `>`, `>=` also?

File: internal/core/unittest/test_expr.cpp
Comment (DragonDriver) [reply internal/core/unittest/test_expr.cpp:585]: Please also add ut for `SegmentInterface.Search` and `SegmentInterface.Retrieve`.
Also, an e2e test demo under `tests/python_client/testcases/` would be perfect.

**Review by longjiquan — CHANGES_REQUESTED** [review]

File: internal/core/src/index/ScalarIndex.h
Comment (avsolatorio) [reply internal/core/src/index/ScalarIndex.h]: RIght! I removed this in the updated version. I followed your advice of processing the data directly using the `elem_func` construct.

File: internal/core/src/query/visitors/ExecExprVisitor.cpp
Comment (avsolatorio) [reply internal/core/src/query/visitors/ExecExprVisitor.cpp]: I implemented an `ExexDataRangeVisitorImpl` that does the filtering of the rows. It passed the tests, but I'd like to ask if the changes I made are what you had in mind, too. :)

File: internal/proxy/plan_parser.go
Comment (avsolatorio) [reply internal/proxy/plan_parser.go]: You're right, I have moved this inside each operation that needs it.

File: internal/core/src/query/visitors/ExecExprVisitor.cpp
Comment (DragonDriver) [reply internal/core/src/query/visitors/ExecExprVisitor.cpp]: Yes! Well-matched what I wanted. thx.

**Comment by DragonDriver** [comment]
@avsolatorio lgtm, also you can complete some details in later pr.

File: internal/core/src/query/visitors/ExecExprVisitor.cpp
Comment (czs007) [inline internal/core/src/query/visitors/ExecExprVisitor.cpp:198]: Please provide more detailed exception assertion information.

Comment (avsolatorio) [reply internal/core/src/query/visitors/ExecExprVisitor.cpp]: Hello @czs007, I have just added a more verbose exception message for this. Thank you!

File: internal/core/src/query/visitors/ExecExprVisitor.cpp
Comment (avsolatorio) [reply internal/core/src/query/visitors/ExecExprVisitor.cpp:267]: Yes, I am thinking of adding these additional comparison operators in a separate PR. I will need to think more about how to better integrate these into the overall architecture of the parser. And potentially refactor the current "==" and "!=" handling, if needed. I hope this is fine. :)

File: internal/proxy/plan_parser.go
Comment (avsolatorio) [reply internal/proxy/plan_parser.go:689]: Ah, I decided to implement the `FunctionNode` since I find it to aptly represent what the feature wanted to accomplish, i.e., to analyze the expression as a function. Nonetheless, I will think through how to better design this and potentially implement the `BinaryArithCmpExpr`  in the next PR if it serves to be a convenient choice when expanding the functionality to support other comparison operators. Also, I remember that I ended up doing the `FunctionNode` transform because, originally, the optimizer will raise an exception if an arithmetic operator has an `IdentifierNode` as its operands. So I had to handle it at the optimizer. 😄

**Comment by avsolatorio** [comment]
> @avsolatorio lgtm, also you can complete some details in later pr.

@DragonDriver, I'll surely create a separate PR for expanding the functionality and also adding more tests. Thank you! 😄

**Comment by DragonDriver** [comment]
rerun ut

**Comment by avsolatorio** [comment]
Hi @DragonDriver , thanks for rerunning the test! I see that it failed again. I'm not sure if the failing test is one of those that intermittently fail or if the failure was caused by my changes. wdyt? 🤔

**Comment by DragonDriver** [comment]
@avsolatorio The ut failure is not caused by your changes. We're fixing it.

**Comment by DragonDriver** [comment]
@avsolatorio The fix of ut has already been merged. You can try rebasing the latest branch to trigger the github action again.

**Comment by avsolatorio** [comment]
@DragonDriver thanks for the heads up, it's passing now. 😄

**Comment by avsolatorio** [comment]
Hello @DragonDriver , the workflows are still failing. Anything that I can do on my end? Also, once the build succeeds, will this PR be merged already? Thank you! cc: @xiaofan-luan

**Comment by DragonDriver** [comment]
rerun ut

**Comment by czs007** [comment]
@avsolatorio good job!  Thanks!
/approve
/lgtm

**Comment by avsolatorio** [comment]
Oh, cool! Thank you, everyone! 😄

**Comment by xiaofan-luan** [comment]
> Oh, cool! Thank you, everyone! 😄

Thanks for the contribution my man, impressive patch~

**Pull request merged**
